Update meta-programming docs (#6261)
lampepfl/dotty: docs/docs/reference/metaprogramming/inline.md

---
layout: doc-page
title: Inline
---

## Inline (whitebox/blackbox)

`inline` is a new [soft modifier](../soft-modifier.html) that guarantees that a
definition will be inlined at the point of use. Example:

```scala
object Config {
  inline val logging = false
}

object Logger {

  private var indent = 0

  inline def log[T](msg: => String)(op: => T): T =
    if (Config.logging) {
      println(s"${"  " * indent}start $msg")
      indent += 1

>       val result = op

Reviewer: Maybe use the safer `val result = try op finally indent -= 1` idiom here.

Author: I think it is sufficient for the exposition inlining

case simply `op` (2)). By-name parameters of the inline method correspond to
`def` bindings whereas by-value parameters correspond to `val` bindings. So if
`log` was defined like this:

```scala
inline def log[T](msg: String)(op: => T): T = ...
```

we'd get

```scala
val msg = s"factorial($n)"
```

instead. This behavior is designed so that calling an inline method is
semantically the same as calling a normal method: By-value arguments
are evaluated before the call whereas by-name arguments are evaluated
each time they are referenced. As a consequence, it is often
preferable to make arguments of inline methods by-name in order to
avoid unnecessary evaluations. Additionally, in the code above, our
goal is to print the result after the evaluation of `op`. Imagine, if we were
printing the duration of the evaluation between the two prints.

For instance, here is how we can define a zero-overhead `foreach` method
that translates into a straightforward while loop without any indirection or
overhead:

```scala
inline def foreach(op: => Int => Unit): Unit = {
  var i = from
  while (i < end) {
    op(i)
    i += 1
  }
}
```

> By contrast, if `op` is a call-by-value parameter, it would be evaluated separately as a closure.

Reviewer 2: The reason why we do not generate the closure is that when we only refer once to a by-name parameter once we can inline the code directly. Which then allows us to beta reduce the call. But if we use it more than one the def will be created to avoid duplication of code. All this is not clear at this point.

For example:

```scala
inline def power(x: Double, inline n: Int): Double
```

### Relationship to @inline

Scala also defines a `@inline` annotation which is used as a hint
for the backend to inline. The `inline` modifier is a more powerful
option: Expansion is guaranteed instead of best effort,
it happens in the frontend instead of in the backend, and it also applies
to recursive methods.

To cross compile between both Dotty and Scalac, we introduce a new `@forceInline`
annotation which is equivalent to the new `inline` modifier. Note that
Scala 2 ignores the `@forceInline` annotation, so one must use both
annotations to guarantee inlining for Dotty and at the same time hint inlining
for Scala 2 (i.e. `@forceInline @inline`).

### Evaluation Rules

As you noticed by the examples above a lambda of the form

`((x_1, ..., x_n) => B)(E_1, ..., E_n)` is rewritten to:

```
{ val/def x_1 = E_1
  ...
  val/def x_n = E_n
  B
}
```

where vals are used for value parameter bindings and defs are used for by-name
parameter bindings. If an argument `E_i` is a simple variable reference `y`, the
corresponding binding is omitted and `y` is used instead of `x_i` in `B`.

> If a inline modifier is given for parameters, corresponding arguments must be

Reviewer 2: ```suggestion
If a `inline` modifier is given for parameters, corresponding arguments must be
```

pure expressions of constant type.

#### The definition of constant expression

Right-hand sides of inline values and of arguments for inline parameters must be
constant expressions in the sense defined by the [SLS §
6.24](https://www.scala-lang.org/files/archive/spec/2.12/06-expressions.html#constant-expressions),

> including "platform-specific" extensions such as constant folding of pure

Reviewer 2: ```suggestion
including _platform-specific_ extensions such as constant folding of pure
```

numeric computations.

#### Implicit Match

> 

Reviewer 2: I would add at least a single sentence describing it.

https://github.com/lampepfl/dotty/pull/5392/files

### Specializing Inline (Whitebox)

> Inline methods support the `<: T` return syntax. This means that the return type

Reviewer 2: ```suggestion
Inline methods support the ` <: T` return type syntax. This means that the return type
```

of the inline method is going to be specialized to a more precise type upon
expansion.

Consider the example below where the inline method `choose` can return an object
of any of the two dynamic types. The subtype relationship is `B <: A`. Since we

> use the specializing inline syntax, the static types of the vals are inferred

Reviewer 2: ```suggestion
use the specializing inline syntax, the static types of the `val`s are inferred
```

> accordingly. Consequently, calling `meth` on `obj2` is a compile-time error.

Reviewer 2: ```suggestion
accordingly. Consequently, calling `meth` on `obj2` is not a compile-time error as `obj2` will be of type `B`.
```

```scala
class A
class B extends A {
  def meth() = true
}

inline def choose(b: Boolean) <: A = {

>   if(b) new A()

Reviewer 2: ```suggestion
  if (b) new A()
```

  else new B()
}

val obj1 = choose(true)  // static type is A
val obj2 = choose(false) // static type is B

obj1.meth() // compile-time error
obj2.meth() // OK
```

> ```scala

Reviewer 2: Add a small description

inline def zero() <: Int = 0

final val one: 1 = zero() + 1
```

#### Inline Match

A `match` expression in the body of an `inline` method definition may be
prefixed by the `inline` modifier. If there is enough static information to
unambiguously take a branch, the expression is reduced to that branch and the
type of the result is taken. The example below defines an inline method with a
single inline match expression that picks a case based on its static type:

```scala
inline def g(x: Any) <: Any = inline x match {
  case x: String => (x, x) // Tuple2[String, String](x, x)
  case x: Double => x
}

> g(1.0d): 1.0d

Reviewer 2: ```suggestion
g(1.0d) // Has type 1.0d which is a subtype of Double
```

> val t: (String, String) = g("test")

Reviewer 2: ```suggestion
g("test") // Has type (String, String)
```

```

The scrutinee `x` is examined statically and the inline match is reduced
accordingly returning the corresponding value (with the type specialized due to
the `<:` in the return type). This example performs a simple type test over the
scrutinee. The type can have a richer structure like the simple ADT below.
`toInt` matches the structure of a number in Church-encoding and _computes_ the
corresponding integer.

```scala
trait Nat
case object Zero extends Nat
case class Succ[N <: Nat](n: N) extends Nat

inline def toInt(n: Nat) <: Int = inline n match {
  case Zero => 0
  case Succ(n1) => toInt(n1) + 1
}

> final val two = toInt(Succ(Succ(Zero)))

Reviewer 2: ```suggestion
final val natTwo = toInt(Succ(Succ(Zero)))
```

> val two_ : 2 = two

Reviewer 2: ```suggestion
val intTwo: 2 = natTwo
```

```

> `two` is inferred to have the singleton type 2.

Reviewer 2: ```suggestion
`natTwo` is inferred to have the singleton type 2.
```

> #### scala.compiletime._

Reviewer 2: We should also mention `scala.compiletime.error`

> 

Reviewer 2: Add a sentence describing what is contained in `scala.compiletime`.

##### Erased Value

We have seen so far inline methods that take terms (tuples and integers) as
parameters. What if we want to base case distinctions on types instead? For
instance, one would like to be able to write a function `defaultValue`, that,
given a type `T` returns optionally the default value of `T`, if it exists. In
fact, we can already express this using rewrite match expressions and a simple
helper function, `scala.compiletime.erasedValue`, which is defined as follows:

```scala
erased def erasedValue[T]: T = ???
```

The `erasedValue` function _pretends_ to return a value of its type argument
`T`. In fact, it would always raise a `NotImplementedError` exception when
called. But the function can in fact never be called, since it is declared
`erased`, so can be only used a compile-time during type checking.

Using `erasedValue`, we can then define `defaultValue` as follows:

```scala

> inline def defaultValue[T] = inline erasedValue[T] match {

Reviewer 2: Put `defaultValue` in another section before `erasedValue`. Describe what it does but do not show the implementation. You can keep this code as an example of how to use `erasedValue` to define `defaultValue`

Author: `defaultValue` is merely an example, not part of the `compiletime` package. Does it deserve a section?

  case _: Byte => Some(0: Byte)
  case _: Char => Some(0: Char)
  case _: Short => Some(0: Short)
  case _: Int => Some(0)
  case _: Long => Some(0L)
  case _: Float => Some(0.0f)
  case _: Double => Some(0.0d)
  case _: Boolean => Some(false)
  case _: Unit => Some(())
  case _: t >: Null => Some(null)
  case _ => None
}
```

Then:
```scala
defaultValue[Int] = Some(0)
defaultValue[Boolean] = Some(false)
defaultValue[String | Null] = Some(null)
defaultValue[AnyVal] = None
```

As another example, consider the type-level version of `toNat` above: given a
_type_ representing a Peano number, return the integer _value_ corresponding to
it. Here's how this can be defined:

```scala
inline def toInt[N <: Nat] <: Int = inline scala.compiletime.erasedValue[N] match {
  case _: Zero => 0
  case _: Succ[n] => toIntT[n] + 1
}

final val two = toInt[Succ[Succ[Zero]]]
```

`erasedValue` is an `erased` method so it cannot be used and has no runtime
behavior. Since `toInt` performs static checks over the static type of `N` we
can safely use it to scrutinize its return type (`S[S[Z]]` in this case).

> ##### Const Value & Const Value Opt

Reviewer 2: This must go before erased values

Reviewer 2: You never showed what is `constValueOpt`

`constvalue` is a function that produces the constant value represented by a
type.

```scala
inline def toIntC[N] <: Int =
  inline scala.compiletime.constValue[N] match {
    case 0 => 0

>     case _: scala.compiletime.S[n1] => 1 + toIntC[n1]

Reviewer 2: We also need to explain what is `scala.compiletime.S`
